src/connectors/query-rules/connectQueryRules.ts

```diff
@@ -0,0 +1,237 @@
+
+export type QueryRulesWidgetFactory<T> = WidgetFactory<
+  QueryRulesConnectorParams & T
+>;
+
+export type QueryRulesConnector = {
+  <T>(
+    render: QueryRulesRenderer<T>,
+    unmount?: () => void
+  ): QueryRulesWidgetFactory<T>;
+};
+
+// A context rule must consist only of alphanumeric characters, hyphens, and underscores.
+// See https://www.algolia.com/doc/guides/managing-results/refine-results/merchandising-and-promoting/in-depth/implementing-query-rules/#context
+function escapeRuleContext(ruleName: string) {
+  return ruleName.replace(/[^a-z0-9-_]+/gi, '_');
+}
+
+function getRuleContextsFromTrackedFilters({
+  helper,
+  trackedFilters,
+}: {
+  helper: Helper;
+  trackedFilters: ParamTrackedFilters;
+}) {
+  const ruleContexts = Object.keys(trackedFilters).reduce(
+    (facets: string[], facetName) => {
+      const getFacetValues = trackedFilters[facetName];
+      const facetRefinements: string[] = helper
+        .getRefinements(facetName)
+        .map((refinement: FacetRefinement) => refinement.value);
+      const facetValues = getFacetValues(facetRefinements);
+
+      return [
+        ...facets,
+        ...facetRefinements
+          .filter(
+            facetRefinement =>
+              facetValues.findIndex(
+                facetValue => facetValue === facetRefinement
+              ) !== -1
+          )
+          .map(facetValue =>
+            escapeRuleContext(`ais-${facetName}-${facetValue}`)
+          ),
+      ];
+    },
+    []
+  );
+
+  return ruleContexts;
+}
+
+function prepareRuleContexts({
+  initialRuleContexts,
+  newRuleContexts,
+  transformRuleContexts,
+}: {
+  initialRuleContexts: string[];
+  newRuleContexts: string[];
+  transformRuleContexts: ParamTransformRuleContexts;
+}) {
+  const allRuleContexts = [...initialRuleContexts, ...newRuleContexts];
+
+  warning(
+    allRuleContexts.length <= 10,
+    `
+The maximum number of \`ruleContexts\` is 10. They have been sliced to that limit.
+Consider using \`transformRuleContexts\` to minimize the number of rules sent to Algolia.
+`
+  );
+
+  const ruleContexts = transformRuleContexts(allRuleContexts).slice(0, 10);
+
+  return ruleContexts;
+}
+
+function applyRuleContexts({ helper, ruleContexts }) {
+  const previousRuleContexts = helper.getQueryParameter('ruleContexts');
+
+  if (!isEqual(previousRuleContexts, ruleContexts)) {
+    helper.setQueryParameter('ruleContexts', ruleContexts).search();
+  }
+}
+
+const connectQueryRules: QueryRulesConnector = (render, unmount = noop) => {
+  checkRendering(render, withUsage());
+
+  // We store the initial rule contexts applied before creating the widget
+  // so that we do not override them with the rules created from `trackedFilters`.
+  let initialRuleContexts: string[] = [];
+
+  // We track the added rule contexts to remove them when the widget is disposed.
+  let addedRuleContexts: string[] = [];
+
+  return widgetParams => {
+    const {
+      trackedFilters = {} as ParamTrackedFilters,
```

Comment:
Unsure about this casting pattern. It's a way to tell TypeScript that the widget parameters are not `undefined` at this step.

Reply:
It's not typed at this step because TypeScript doesn't infer the `QueryRulesConnectorParams` types.

![image](https://user-images.githubusercontent.com/6137112/54200478-c8d59080-44cb-11e9-9ec4-6645abe85964.png)

Reply:
That's unexpected, it should be inferred I think.